test/transaction/large_transaction_test.cpp

```diff
@@ -0,0 +1,286 @@
+  }
+
+  void UpdateSnapshot(RandomWorkloadTransaction *txn,
+                      TableSnapshot *curr, const TableSnapshot &before) {
+    for (auto &entry : before)
+      curr->emplace(entry.first, CopyTuple(entry.second));
+    for (auto &update : txn->updates_) {
+      // TODO(Tianyu): Can be smarter about copies
+      storage::ProjectedRow *new_version = (*curr)[update.first];
+      storage::StorageUtil::ApplyDelta(layout_, *update.second, new_version);
+    }
+  }
+
+  // This returned value will contain memory that has to be freed manually
+  VersionedSnapshots ReconstructVersionedTable(std::vector<RandomWorkloadTransaction *> *txns) {
+    VersionedSnapshots result;
+    // empty starting version
+    TableSnapshot *prev = &(result.emplace(timestamp_t(0), TableSnapshot()).first->second);
+    // populate with initial image of the table
+    for (auto &entry : initial_txn_.inserts_)
+      (*prev)[entry.first] = CopyTuple(entry.second);
+
+    for (RandomWorkloadTransaction *txn : *txns) {
+      auto ret = result.emplace(txn->commit_time_, TableSnapshot());
+      UpdateSnapshot(txn, &(ret.first->second), *prev);
+      prev = &(ret.first->second);
+    }
+    return result;
+  }
+
+  void CheckTransactionReadCorrect(RandomWorkloadTransaction *txn,
+                                   const VersionedSnapshots &snapshots) {
+    timestamp_t start_time = txn->txn_->StartTime();
+    // this version is the most recent future update
+    auto ret = snapshots.upper_bound(start_time);
+    // Go to the version visible to this txn
+    --ret;
+    timestamp_t version_timestamp = ret->first;
+    const TableSnapshot &before_snapshot = ret->second;
+    EXPECT_TRUE(transaction::TransactionUtil::NewerThan(start_time, version_timestamp));
+    for (auto &entry : txn->selects_) {
+      auto it = before_snapshot.find(entry.first);
+      EXPECT_TRUE(StorageTestUtil::ProjectionListEqual(layout_, entry.second, it->second));
+    }
+  }
+
+  uint32_t txn_length_;
+  std::vector<double> update_select_ratio_;
+  std::default_random_engine *generator_;
+  storage::BlockLayout layout_;
+  storage::DataTable table_;
+  transaction::TransactionManager txn_manager_;
+  std::vector<storage::TupleSlot> all_slots_;
+  RandomWorkloadTransaction initial_txn_;
+};
+
+class LargeTransactionTests : public ::testing::Test {
+ public:
+  storage::BlockStore block_store_{1000};
+  common::ObjectPool<transaction::UndoBufferSegment> buffer_pool_{1000};
+  std::default_random_engine generator_;
+};
+
+// This test case generates random update-selects in concurrent transactions on a pre-populated database.
+// Each transaction logs their operations locally. At the end of the run, we can reconstruct the snapshot of
+// a database using the updates at every timstamp, and compares the reads with the reconstructed snapshot versions
+// to make sure they are the same.
+// NOLINTNEXTLINE
+TEST_F(LargeTransactionTests, MixedReadWrite) {
```

Comment:
We'll want to add more test cases (update/select ratio, tuples, etc.) but this is good for now. That can be done next week during the cleanup.